File: coursework/capture_swagger.py
import os
import sys
import time
import urllib.request
import subprocess

def wait_for_service(url="http://127.0.0.1:3000/health", timeout=60):
    start = time.time()
    print(f"Waiting for {url} to become ready...")
    while time.time() - start < timeout:
        try:
            with urllib.request.urlopen(url, timeout=2) as resp:
                if resp.status == 200:
                    print("Service is ready!")
                    return True
        except Exception:
            time.sleep(1)
    print("Timed out waiting for service.")
    return False

def take_screenshot(url, output_path, window_size="1400,1100"):
    cmd = [
        "/etc/profiles/per-user/zerok/bin/chromium",
        "--headless=new",
        "--disable-gpu",
        "--hide-scrollbars",
        f"--window-size={window_size}",
        f"--screenshot={output_path}",
        url
    ]
    print(f"Capturing {url} -> {output_path}")
    res = subprocess.run(cmd, capture_output=True, text=True)
    if os.path.exists(output_path):
        print(f"Successfully captured: {output_path} ({os.path.getsize(output_path)} bytes)")
        return True
    else:
        print(f"Failed to capture. Stderr: {res.stderr}")
        return False

def main():
    if not wait_for_service():
        sys.exit(1)
    
    os.makedirs("/home/zerok/projects/upward/screenshots", exist_ok=True)
    
    # 1. Capture Swagger Main Overview
    take_screenshot(
        "http://127.0.0.1:3000/docs/#/",
        "/home/zerok/projects/upward/screenshots/swagger_overview.png",
        window_size="1400,1200"
    )
    
    # 2. Capture Health endpoint
    take_screenshot(
        "http://127.0.0.1:3000/health",
        "/home/zerok/projects/upward/screenshots/health_endpoint.png",
        window_size="800,400"
    )

if __name__ == "__main__":
    main()

File: coursework/capture_swagger.js
const { spawn } = require('child_process');
const fs = require('fs');

async function run() {
  const chrome = spawn('/etc/profiles/per-user/zerok/bin/chromium', [
    '--headless=new',
    '--remote-debugging-port=9223',
    '--disable-gpu',
    '--window-size=1400,1800',
    'http://127.0.0.1:3000/docs/'
  ]);

  await new Promise(r => setTimeout(r, 2500));

  try {
    const listRes = await fetch('http://127.0.0.1:9223/json');
    const tabs = await listRes.json();
    const pageTab = tabs.find(t => t.type === 'page');
    if (!pageTab) throw new Error('No page tab found');

    const ws = new WebSocket(pageTab.webSocketDebuggerUrl);
    await new Promise(res => ws.onopen = res);

    let id = 1;
    function send(method, params = {}) {
      return new Promise((resolve) => {
        const curId = id++;
        const handler = (evt) => {
          const msg = JSON.parse(evt.data);
          if (msg.id === curId) {
            ws.removeEventListener('message', handler);
            resolve(msg.result);
          }
        };
        ws.addEventListener('message', handler);
        ws.send(JSON.stringify({ id: curId, method, params }));
      });
    }

    await send('Runtime.enable');
    await send('Page.enable');

    // Wait for swagger to render
    await new Promise(r => setTimeout(r, 2000));

    // Expand all endpoints
    await send('Runtime.evaluate', {
      expression: `
        document.querySelectorAll('.opblock-summary').forEach(el => el.click());
      `
    });

    await new Promise(r => setTimeout(r, 1500));

    // Capture screenshot of expanded operations
    const { data: b64Expanded } = await send('Page.captureScreenshot', {
      format: 'png',
      captureBeyondViewport: true
    });

    fs.writeFileSync('/home/zerok/projects/upward/screenshots/swagger_expanded.png', Buffer.from(b64Expanded, 'base64'));
    console.log('Saved swagger_expanded.png successfully!');

    ws.close();
  } finally {
    chrome.kill();
  }
}

run().catch(console.error);
